File: iRank Timer/BlindSet.m
//
//  BlindSet.m
//  iRank Timer
//
//  Created by Luciano Stegun on 2013-09-22.
//  Copyright (c) 2013 Stegun.com. All rights reserved.
//

#import "BlindSet.h"

@implementation BlindSet
@synthesize blindSetId;
@synthesize blindSetIndex;
@synthesize duration;
@synthesize blindSetName;
@synthesize blindLevelList;
@synthesize currentLevel;
@synthesize elapsedSeconds;
@synthesize nextBreakRemain;
@synthesize breaks;
@synthesize levels;
@synthesize isNew;
@synthesize playSound;
@synthesize blindChangeAlert;
@synthesize minuteAlert;
@synthesize doublePrevious;
@synthesize seconds;

-(id)init {
    
    self = [super init];
    
    blindLevelList = [[NSMutableArray alloc] init];
    currentLevel   = 1;
    elapsedSeconds = 0;
    levels         = 0;
    breaks         = 0;
    
    playSound        = YES;
    blindChangeAlert = YES;
    minuteAlert      = YES;
    doublePrevious   = YES;
    
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    
    BlindSet *copy = [[BlindSet alloc] init];
    
    if( copy ){
        
        [copy setBlindSetId:         self.blindSetId];
        [copy setBlindSetIndex:      self.blindSetIndex];
        [copy setElapsedSeconds:     self.elapsedSeconds];
        [copy setCurrentLevel:       self.currentLevel];
        [copy setNextBreakRemain:    self.nextBreakRemain];
        [copy setPlaySound:          self.playSound];
        [copy setBlindChangeAlert:   self.blindChangeAlert];
        [copy setMinuteAlert:        self.minuteAlert];
        [copy setDoublePrevious:     self.doublePrevious];
        [copy setBlindSetName:       self.blindSetName];
        [copy setDuration:           self.duration];
        [copy setBlindLevelList:     [[NSMutableArray alloc] init]];
        [copy setBreaks:             self.breaks];
        [copy setLevels:             self.levels];
        [copy setSeconds:            self.seconds];
        
        for( BlindLevel *blindLevel in self.blindLevelList )
            [copy.blindLevelList addObject:[blindLevel copy]];
    }
    
    return copy;
}

- (void)encodeWithCoder:(NSCoder *)encoder {
    
    [encoder encodeBool:isNew forKey:@"isNew"];
    [encoder encodeInt:blindSetId forKey:@"blindSetId"];
    [encoder encodeInt:blindSetIndex forKey:@"blindSetIndex"];
    [encoder encodeInt:elapsedSeconds forKey:@"elapsedSeconds"];
    [encoder encodeInt:currentLevel forKey:@"currentLevel"];
    [encoder encodeInt:nextBreakRemain forKey:@"nextBreakRemain"];
    [encoder encodeBool:playSound forKey:@"playSound"];
    [encoder encodeBool:blindChangeAlert forKey:@"blindChangeAlert"];
    [encoder encodeBool:minuteAlert forKey:@"minuteAlert"];
    [encoder encodeBool:doublePrevious forKey:@"doublePrevious"];
    [encoder encodeInt:breaks forKey:@"breaks"];
    [encoder encodeInt:levels forKey:@"levels"];
    [encoder encodeInt:seconds forKey:@"seconds"];
    [encoder encodeObject:blindSetName forKey:@"blindSetName"];
    [encoder encodeObject:duration forKey:@"duration"];
    [encoder encodeObject:blindLevelList forKey:@"blindLevelList"];
}

- (id)initWithCoder:(NSCoder *)decoder {
    
    self = [super init];
    
    isNew = NO;
    
    [self setBlindSetId:         [decoder decodeIntForKey:@"blindSetId"]];
    [self setBlindSetIndex:      [decoder decodeIntForKey:@"blindSetIndex"]];
    [self setElapsedSeconds:     [decoder decodeIntForKey:@"elapsedSeconds"]];
    [self setCurrentLevel:       [decoder decodeIntForKey:@"currentLevel"]];
    [self setNextBreakRemain:    [decoder decodeIntForKey:@"nextBreakRemain"]];
    [self setPlaySound:          [decoder decodeBoolForKey:@"playSound"]];
    [self setBlindChangeAlert:   [decoder decodeBoolForKey:@"blindChangeAlert"]];
    [self setMinuteAlert:        [decoder decodeBoolForKey:@"minuteAlert"]];
    [self setDoublePrevious:     [decoder decodeBoolForKey:@"doublePrevious"]];
    [self setBlindSetName:       [decoder decodeObjectForKey:@"blindSetName"]];
    [self setDuration:           [decoder decodeObjectForKey:@"duration"]];
    [self setBlindLevelList:     [decoder decodeObjectForKey:@"blindLevelList"]];

    breaks  = [decoder decodeIntForKey:@"breaks"];
    levels  = [decoder decodeIntForKey:@"levels"];
    seconds = [decoder decodeIntForKey:@"seconds"];
    
    return self;
}

-(void)addBlindLevel:(BlindLevel*)blindLevel {

    if( blindLevel.isBreak )
        breaks++;
    else
        levels++;
    
    seconds += blindLevel.seconds;
    
    [blindLevelList addObject:blindLevel];
}

-(int)blindLevels {
    
    return blindLevelList.count;
}

-(BlindLevel*)blindLevelByNumber:(int)levelNumber {
    
    for(BlindLevel *blindLevel in blindLevelList)
        if( blindLevel.levelNumber==levelNumber )
            return blindLevel;
        
    return nil;
}

-(BlindLevel*)blindLevelByIndex:(int)levelIndex {
    
    if( levelIndex > self.blindLevels-1 ){
        
        currentLevel = self.blindLevels;
        levelIndex   = currentLevel-1;
    }
    
    return (BlindLevel *)[blindLevelList objectAtIndex:levelIndex];
}

-(BlindLevel *)currentBlindLevel {
    
    return [self blindLevelByIndex:currentLevel-1];
}

-(void)updateElapsedTime {
    
    int elapsedTime = 0;
    
    for(BlindLevel *blindLevel in blindLevelList){
        
        elapsedTime += blindLevel.seconds;
        
        blindLevel.elapsedTime = [BlindSet formatTimeString:elapsedTime];
    }
}

-(int)previousLevel {
    
    if( currentLevel > 1 )
        elapsedSeconds -= [[self blindLevelByIndex:currentLevel-2] seconds];

    currentLevel = currentLevel - 1;
    
    if( currentLevel <= 0 )
        currentLevel = 1;
    
    [self updateNextBreakRemain];
    
    return currentLevel;
}

-(int)nextLevel {

    if( currentLevel < self.blindLevels )
        elapsedSeconds += [[self currentBlindLevel] seconds];
    
    currentLevel = currentLevel+1;
    
    if( currentLevel > self.blindLevels )
        currentLevel = self.blindLevels;
    
    [self updateNextBreakRemain];
    
    return currentLevel;
}

-(int)resetLevels {
    
    currentLevel   = 1;
    elapsedSeconds = 0;
    [self updateNextBreakRemain];
    
    return currentLevel;
}

-(NSString *)elapsedtime {
    
    if( currentLevel > 1 ){
        
        BlindLevel *aBlindLevel = [self blindLevelByIndex:currentLevel-2];
        return [aBlindLevel elapsedTime];
    }
    
    return @"00:00:00";
}

+(NSString *)formatTimeString:(int)seconds {
    
    int hours   = floor(seconds/3600);
    seconds    -= (hours*3600);
    int minutes = floor(seconds/60);
    seconds    -= (minutes*60);
    
    return [NSString stringWithFormat:@"%02i:%02i:%02i", hours, minutes, seconds];
}

-(void)updateNextBreakRemain {
    
    int nextBreak   = 0;
    BOOL hasBreak   = NO;
    BOOL isBreakNow = self.currentBlindLevel.isBreak;
    
    int levelNumber = 0;
    for(BlindLevel *blindLevel in blindLevelList){
        
        levelNumber++;
        
        if( levelNumber < currentLevel )
            continue;
        
        if( blindLevel.isBreak && (!isBreakNow || levelNumber > currentLevel) ){
            
            hasBreak = YES;
            break;
        }
        
        nextBreak += blindLevel.seconds;
    }
    
    if( hasBreak )
        nextBreakRemain = nextBreak;
    else
        nextBreakRemain = -1;
}

-(void)resetLevelNumbers {
    
    int durationSeconds = 0;
    int levelNumber     = 0;
    int levelIndex      = 0;
    levels = 0;
    breaks = 0;
    
    for(BlindLevel *blindLevel in blindLevelList){
        
        durationSeconds += blindLevel.seconds;
        
        blindLevel.elapsedTime = [BlindSet formatTimeString:durationSeconds];
        blindLevel.levelIndex  = levelIndex++;
        
        if( blindLevel.isBreak ){
            
            breaks++;
            continue;
        }
        
        levels++;
        
        levelNumber++;
        blindLevel.levelNumber = levelNumber;
    }
    
    self.duration = [BlindSet formatTimeString:durationSeconds];
    seconds       = durationSeconds;
}

+ (NSMutableArray *)loadBlindSetListByXml:(NSString*)xmlFile {
    
    XmlBlindSetParser *blindSetParser = [[XmlBlindSetParser alloc] initXMLParser];
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:xmlFile ofType:@"xml"];
    NSData *xmlData = [NSData dataWithContentsOfFile:filePath];
    
    NSXMLParser *parser = [[NSXMLParser alloc] initWithData:xmlData];
    
    [parser setDelegate:blindSetParser];
    [parser setShouldProcessNamespaces:YES];
    [parser setShouldReportNamespacePrefixes:YES];
    [parser setShouldResolveExternalEntities:NO];
    [parser parse];
    
    return [blindSetParser getBlindSetList];
}

+ (NSMutableArray *)loadArchivedBlindSetList:(BOOL)archive {
    
    NSString *blindSetListPath = [BlindSet blindSetArrayPath];
    
    NSMutableArray *blindSetList;
    NSFileManager *manager = [NSFileManager defaultManager];
    BOOL fileExists = [manager fileExistsAtPath:blindSetListPath];
    
    if( fileExists )
        blindSetList = [NSKeyedUnarchiver unarchiveObjectWithFile:blindSetListPath];
    else{
     
        blindSetList = [BlindSet loadBlindSetListByXml:(LITE_VERSION?@"startupBlindSets-lite":@"startupBlindSets")];
        
        // Se o arquivo ainda não existir
        if( archive )
           [BlindSet archivedBlindSetList: blindSetList];
    }
    
    // Se for a versão free corta a lista deixando apenas o primeiro
    if( LITE_VERSION )
        blindSetList = [NSMutableArray arrayWithObject:blindSetList.firstObject];
    
    return blindSetList;
}

+ (NSString *)blindSetArrayPath {
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    return [documentsDirectory stringByAppendingPathComponent:@"customBlindSetList.data"];
}

+ (void)archivedBlindSetList:(NSMutableArray*)blindSetList {

    NSString *blindSetListPath = [BlindSet blindSetArrayPath];
    [NSKeyedArchiver archiveRootObject:blindSetList toFile:blindSetListPath];
}

-(NSString *)description {
    
    return [NSString stringWithFormat:@"#%i (%i): %@, L:%i B:%i D:%@", blindSetId, blindSetIndex, blindSetName, levels, breaks, duration];
}

@end

File: irank_timer/trunk/iRank Timer/BlindSet.swift
//
//  BlindSet.swift
//  iRank Timer
//
//  Created by Luciano Stegun on 05/02/15.
//  Copyright (c) 2015 Stegun.com. All rights reserved.
//

import UIKit
import Foundation

@objc(BlindSet)
class BlindSet : NSObject, NSCoding {
    
    class func appDelegate() -> AppDelegate {
        
        return UIApplication.sharedApplication().delegate as! AppDelegate;
    }
    
    var isNew : Bool = false;
    var blindSetId : Int;
    var blindSetIndex : Int = 0;
    var elapsedSeconds : Int = 0;
    var currentElapsedSeconds : Int = 0;
    var currentLevelIndex : Int = 0;
    var currentLevel : Int = 0; // Legado da versão 1.0
    var nextBreakRemain : Int = 0;
    var playSound : Bool = true;
    var blindChangeAlert : Bool = true;
    var lastMinuteAlert : Bool = true;
    var minuteAlert : Bool = true; // Legado da versão 1.0
    var doublePrevious : Bool = true;
    var breaks : Int = 0;
    var levels : Int = 0;
    var seconds : Int = 0;
    var blindSetName : String = "New blind set";
    var duration : String = "00:00:00";
    var blindLevelList : Array<BlindLevel> = [];
    
    override init(){
        
        self.blindSetId        = 0;
        self.currentLevelIndex = 0;
        self.levels            = 0;
    }
    
    required init(coder aDecoder: NSCoder) {
        
        self.isNew = false;
        
        self.blindSetId        = 0;
        self.currentLevelIndex = 0;
        self.levels            = 0;
        
        self.blindSetId            = Int(aDecoder.decodeIntForKey("blindSetId"));
        self.blindSetIndex         = Int(aDecoder.decodeIntForKey("blindSetIndex"));
        self.elapsedSeconds        = Int(aDecoder.decodeIntForKey("elapsedSeconds"));
        self.currentElapsedSeconds = Int(aDecoder.decodeIntForKey("currentElapsedSeconds"));
        self.currentLevelIndex     = Int(aDecoder.decodeIntForKey("currentLevelIndex"));
        self.currentLevel          = Int(aDecoder.decodeIntForKey("currentLevelIndex"));
        self.nextBreakRemain       = Int(aDecoder.decodeIntForKey("nextBreakRemain"));
        self.playSound             = aDecoder.decodeBoolForKey("playSound");
        self.blindChangeAlert      = aDecoder.decodeBoolForKey("blindChangeAlert");
        self.minuteAlert           = aDecoder.decodeBoolForKey("minuteAlert");
        self.lastMinuteAlert       = aDecoder.decodeBoolForKey("lastMinuteAlert");
        self.doublePrevious        = aDecoder.decodeBoolForKey("doublePrevious");
        self.blindSetName          = aDecoder.decodeObjectForKey("blindSetName") as! String;
        self.duration              = aDecoder.decodeObjectForKey("duration") as! String;
        self.blindLevelList        = aDecoder.decodeObjectForKey("blindLevelList") as! Array<BlindLevel>;
        self.breaks                = Int(aDecoder.decodeIntForKey("breaks"));
        self.levels                = Int(aDecoder.decodeIntForKey("levels"));
        self.seconds               = Int(aDecoder.decodeIntForKey("seconds"));
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        
        aCoder.encodeBool(self.isNew, forKey:"isNew");
        aCoder.encodeInt(Int32(self.blindSetId), forKey:"blindSetId");
        aCoder.encodeInt(Int32(self.blindSetIndex), forKey:"blindSetIndex");
        aCoder.encodeInt(Int32(self.currentElapsedSeconds), forKey:"currentElapsedSeconds");
        aCoder.encodeInt(Int32(self.elapsedSeconds), forKey:"elapsedSeconds");
        aCoder.encodeInt(Int32(self.currentLevelIndex), forKey:"currentLevelIndex");
        aCoder.encodeInt(Int32(self.nextBreakRemain), forKey:"nextBreakRemain");
        aCoder.encodeBool(self.playSound, forKey:"playSound");
        aCoder.encodeBool(self.blindChangeAlert, forKey:"blindChangeAlert");
        aCoder.encodeBool(self.minuteAlert, forKey:"minuteAlert");
        aCoder.encodeBool(self.lastMinuteAlert, forKey:"lastMinuteAlert");
        aCoder.encodeBool(self.doublePrevious, forKey:"doublePrevious");
        aCoder.encodeInt(Int32(self.breaks), forKey:"breaks");
        aCoder.encodeInt(Int32(self.levels), forKey:"levels");
        aCoder.encodeInt(Int32(self.seconds), forKey:"seconds");
        aCoder.encodeObject(self.blindSetName, forKey:"blindSetName");
        aCoder.encodeObject(self.duration, forKey:"duration");
        aCoder.encodeObject(self.blindLevelList, forKey:"blindLevelList");
    }
    
    override func copy() -> AnyObject {
        
        var blindSet = BlindSet();
        
        blindSet.isNew                 = self.isNew;
        blindSet.blindSetId            = self.blindSetId;
        blindSet.blindSetIndex         = self.blindSetIndex;
        blindSet.currentElapsedSeconds = self.currentElapsedSeconds;
        blindSet.elapsedSeconds        = self.elapsedSeconds;
        blindSet.currentLevelIndex     = self.currentLevelIndex;
        blindSet.nextBreakRemain       = self.nextBreakRemain;
        blindSet.playSound             = self.playSound;
        blindSet.blindChangeAlert      = self.blindChangeAlert;
        blindSet.minuteAlert           = self.minuteAlert;
        blindSet.lastMinuteAlert       = self.lastMinuteAlert;
        blindSet.doublePrevious        = self.doublePrevious;
        blindSet.breaks                = self.breaks;
        blindSet.levels                = self.levels;
        blindSet.seconds               = self.seconds;
        blindSet.blindSetName          = self.blindSetName;
        blindSet.duration              = self.duration;
        blindSet.blindLevelList        = [];
        
        for blindLevel in self.blindLevelList {
            
            blindSet.blindLevelList.append(blindLevel.copy() as! BlindLevel);
        }
        
        return blindSet;
    }
    
    func addBlindLevel(blindLevel : BlindLevel){
        
        self.blindLevelList.append(blindLevel);
        self.updateMetadata();
    }
    
    func updateElapsedSeconds(){
        
        var elapsedTime : Int = 0;
        
        if( self.currentLevelIndex > self.blindLevelList.count-1 ){
            
            self.currentLevelIndex = self.blindLevelList.count-1;
        }
        
        for( var i=0; i < self.currentLevelIndex; i++ ){
            
            elapsedTime += self.blindLevelList[i].duration;
        }
        
        self.elapsedSeconds = elapsedTime;
    }
    
    func getElapsedSeconds(levelIndex : Int) -> Int {
        
        var elapsedTime : Int = 0;
        
        for( var i=0; i < levelIndex; i++ ){
            
            elapsedTime += self.blindLevelList[i].duration;
        }
        
        return elapsedTime;
    }
    
    func updateMetadata(){
        
        var levels   = 0;
        var breaks   = 0;
        var duration = 0;
        
        for( var i = 0; i < self.blindLevelList.count; i++ ){
            
            self.blindLevelList[i].levelIndex = i;
            duration                          = duration + blindLevelList[i].duration;
            
            if( self.blindLevelList[i].isBreak == true ){
                
                ++breaks;
            }else{
                
                ++levels;
                self.blindLevelList[i].levelNumber = levels;
            }
        }
        
        self.levels = levels;
        self.breaks = breaks;
        
        updateElapsedSeconds();
        updateNextBreak();
        
        self.duration = Util.formatTimeString(Float(duration)) as! String;
    }
    
    func updateNextBreak(){
        
        var remainingTime = 0;
        var hasBreak      = false;
        
        for( var i = self.currentLevelIndex; i < self.blindLevelList.count; i++ ){
            
            if( self.blindLevelList[i].isBreak == true && i > self.currentLevelIndex ){
                
                hasBreak = true;
                break;
            }
            
            remainingTime += self.blindLevelList[i].duration;
        }
        
        if( hasBreak ){
            
            self.nextBreakRemain = remainingTime;
        }else{
            
            self.nextBreakRemain = 0;
        }
    }
    
    class func findBlindSetListByXml(fileName : String, blindSetId : Int) -> BlindSet{
        
        var blindSetParser : XmlBlindSetParser = XmlBlindSetParser();
        blindSetParser.blindSetId = blindSetId;
        
        var filePath : String = NSBundle.mainBundle().pathForResource(fileName, ofType: "xml")!;
        var xmlData : NSData = NSData(contentsOfFile: filePath)!;
        
        var parser : NSXMLParser = NSXMLParser(data: xmlData);
        
        parser.delegate = blindSetParser
        parser.shouldProcessNamespaces = true;
        parser.shouldReportNamespacePrefixes = true;
        parser.shouldResolveExternalEntities = false;
        parser.parse();
        
        return blindSetParser.blindSet;
    }
    
    class func loadBlindSetListByXml(fileName : String) -> Array<BlindSet>{
        
        var blindSetParser : XmlBlindSetParser = XmlBlindSetParser();
        
        var filePath : String = NSBundle.mainBundle().pathForResource(fileName, ofType: "xml")!;
        var xmlData : NSData = NSData(contentsOfFile: filePath)!;
        
        var parser : NSXMLParser = NSXMLParser(data: xmlData);
        
        parser.delegate = blindSetParser
        parser.shouldProcessNamespaces = true;
        parser.shouldReportNamespacePrefixes = true;
        parser.shouldResolveExternalEntities = false;
        parser.parse();
        
        return blindSetParser.getBlindSetList();
    }
    
    class func archiveBlindSetList(blindSetList : Array<BlindSet>){
        
        var blindSetListPath = BlindSet.getBlindSetArchivePath();
        var success : Bool = NSKeyedArchiver.archiveRootObject(blindSetList, toFile: blindSetListPath);

        if( success && BlindSet.appDelegate().enableBackup ){
            
            BlindSet.saveiCloudBackup();
        }
    }
    
    class func convertArchivedBlindSetList(){
        
        var paths : Array = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true) as Array;
        var documentsDirectory : String = paths[0] as! String;

        var blindSetListPath = "\(documentsDirectory)/customBlindSetList.data";
        
//        println("blindSetListPath: \(blindSetListPath)");
        
        var manager : NSFileManager = NSFileManager.defaultManager();
        var fileExists = manager.fileExistsAtPath(blindSetListPath);
        
        if( !fileExists ){
            
//            println("Arquivo não existe");
            return;
        }
        
        var blindSetOldList : Array<BlindSet> = NSKeyedUnarchiver.unarchiveObjectWithFile(blindSetListPath) as! Array<BlindSet>;
        
        var blindSetList : Array<BlindSet> = [];
        
        for blindSetOld : BlindSet in blindSetOldList {
            
            var blindSet : BlindSet = blindSetOld.copy() as! BlindSet;
            blindSet.currentLevelIndex = blindSetOld.currentLevel;
            blindSet.lastMinuteAlert   = blindSetOld.minuteAlert;

            for( var i = blindSet.currentLevelIndex; i < blindSet.levels; i++ ){
                
                blindSet.blindLevelList[i].duration        = blindSet.blindLevelList[i].duration * 60;
            }
            
            blindSet.updateMetadata();
            
            blindSetList.append(blindSet);
        }
        
        BlindSet.archiveBlindSetList(blindSetList);
    }
    
    class func loadArchivedBlindSetList(archive : Bool) -> Array<BlindSet> {
        
        var blindSetListPath = BlindSet.getBlindSetArchivePath();
        
        var blindSetList : Array<BlindSet> = [];
        var manager : NSFileManager = NSFileManager.defaultManager();
        var fileExists = manager.fileExistsAtPath(blindSetListPath);
        
        if( fileExists ){
            
            blindSetList = NSKeyedUnarchiver.unarchiveObjectWithFile(blindSetListPath) as! Array<BlindSet>;
        }else{

            blindSetList = BlindSet.loadBlindSetListByXml("startupBlindSets");
            
            if( archive ){
            
                BlindSet.archiveBlindSetList(blindSetList);
            }
        }

        return blindSetList;
    }
    
    class func loadArchivedBlindSetList(documentUrl : NSURL, archive: Bool) -> Array<BlindSet>! {
        
        var blindSetListPath : String = documentUrl.path!;
        
        var blindSetList : Array<BlindSet> = [];
        var manager : NSFileManager = NSFileManager.defaultManager();
        var fileExists = manager.fileExistsAtPath(blindSetListPath);
        
        if( fileExists ){
            
            blindSetList = NSKeyedUnarchiver.unarchiveObjectWithFile(blindSetListPath) as! Array<BlindSet>;
            BlindSet.archiveBlindSetList(blindSetList);
            
            return blindSetList;
        }
        
        return nil;
    }
    
    class func getBlindSetArchivePath() -> String {
        
        var paths : Array = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true) as Array;
        var documentsDirectory : String = paths[0] as! String;
        
        return "\(documentsDirectory)/archivedBlindSetList.data";
    }
    
    class func saveiCloudBackup(){
        
        println("Saving data to iCloud: ")
        
        if( !appDelegate().needBackup ){
        
            println("No backup needed")
            return;
        }
        
        var manager : NSFileManager = NSFileManager.defaultManager();
        
        var ubiq : NSURL!;
        ubiq = manager.URLForUbiquityContainerIdentifier(nil);
        
        println("ubiq: \(ubiq)\n-----------------------------------------------------------");
       
        if( ubiq == nil ){
            
            println("Backup no iCloud não habilitado");
            return;
        }
        
        var theError : NSError?;
        var fileExists : Bool   = false;
        var filePath            = BlindSet.getBlindSetArchivePath();
        var backupName : String = "archivedBlindSetList.data";
        
        fileExists = manager.fileExistsAtPath(filePath);
        
        if( !fileExists ){
            
            println("Arquivo fonte para backup no iCloud não existe");
            return;
        }else{

            var deviceName   = appDelegate().device.name;
            var blindSetList = BlindSet.loadArchivedBlindSetList(false);
            let dateFormatter = NSDateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd";
            var backupDate   = dateFormatter.stringFromDate(NSDate());

            backupName = "Backup \(backupDate) "+NSLocalizedString("with", comment:"")+" \(blindSetList.count) "+NSLocalizedString("blind set", comment:"")+"\(Util.getPlural(blindSetList.count)) @ \(deviceName)";
            
            var plainData : NSData = backupName.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: true)!;
            
            backupName = plainData.base64EncodedStringWithOptions(NSDataBase64EncodingOptions.allZeros);
            backupName = "\(backupName).data";
            
            println("backupName: \(backupName)");
        }
        
        var sourceUrl      = NSURL(fileURLWithPath: filePath);
        var destinationUrl = ubiq.URLByAppendingPathComponent("Documents", isDirectory: true).URLByAppendingPathComponent(backupName);
        
//        println("destinationUrl: \(destinationUrl)");
        
        fileExists = manager.fileExistsAtPath(destinationUrl.path!);
        
        if( fileExists ){
            
            manager.removeItemAtURL(destinationUrl, error: &theError)
            
            if( theError != nil ){
                
                println("Error removing file to prepare iCloud Backup");
                return
            }
        }
        
//        println("sourceUrl: \(sourceUrl)\n-----------------------------------------------------------");
        
        manager.setUbiquitous(false, itemAtURL: sourceUrl!, destinationURL: destinationUrl, error: &theError);
//        manager.copyItemAtURL(destinationUrl, toURL: sourceUrl!, error: nil)
        
        if( theError != nil ){
            
            println("theError: \(theError)\n-----------------------------------------------------------");
        }else{
            
            appDelegate().needBackup = false;
            println("Success sent to iCloud backup");
        }
    }
    
    class func checkiCloudBackupAvailability() -> Bool {
 
//        println("Checking iCloud backup availability...");
        
        var ubiq : NSURL!;
        ubiq = NSFileManager.defaultManager().URLForUbiquityContainerIdentifier(nil);
        
        if( ubiq == nil ){
            
//            println("iCloud backup is disabled")
            return false;
        }
        
//        println("iCloud backup is enabled");
        
        return true;
    }
}